### PR TITLE
 :sparkles: Handle variable definition conflicts for external variables

### PR DESCRIPTION
**Pull request by killianmuldoon**
Add handling for conflicting definitions in variables defined in the ClusterClass and external patches. This change includes:
- Comparing variables and signalling which variables conflict in the ClusterClass status
- Validation and Defaulting of variables based on the appropriate definiton where conflicts exist 
- Validating that Cluster's define a `DefinitionFrom` field when there are conflicting variables

Part of #7985

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
/hold

This PR contains a commit from #8076 which should be reviewed and merged first.

**Comment by killianmuldoon** [comment]
This is probably ready for a round of review @sbueringer

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 1]
```diff
@@ -292,6 +304,22 @@ func statusVariableFromClusterClassVariable(variable clusterv1.ClusterClassVaria
...
+	combinedVariable := existingVariable.DeepCopy()
+	newVariableDefinition := clusterv1.ClusterClassStatusVariableDefinition{
+		From:     from,
+		Required: variable.Required,
+		Schema:   variable.Schema,
+	}
+	combinedVariable.Definitions = append(existingVariable.Definitions, newVariableDefinition)
+
+	// If the new definition is different from any existing definition, set DefinitionsConflict to true.
+	if !reflect.DeepEqual(combinedVariable.Definitions[0], newVariableDefinition) {
```
Comment (sbueringer) [inline internal/controllers/clusterclass/clusterclass_controller.go:317]:
```suggestion
	if !combinedVariable.DefinitionsConflict && !reflect.DeepEqual(combinedVariable.Definitions[0], newVariableDefinition) {
```

nit: Small optimization, just saves us a bunch of reflection once we have a conflict in a variable

The following would also achieve the same, but seems relatively hard to read to me:
```go
	combinedVariable.DefinitionsConflict = combinedVariable.DefinitionsConflict ||
		!reflect.DeepEqual(combinedVariable.Definitions[0], newVariableDefinition)
```

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package clusterclass
 import (
```
Comment (sbueringer) [inline internal/controllers/clusterclass/clusterclass_controller.go:19]: I assume unit tests for the new additional behavior are TODO
(I'll assume the same in case I see other untested code in other files)

Comment (killianmuldoon) [reply internal/controllers/clusterclass/clusterclass_controller.go:19]: I think this is the only place where there should be untested code in this PR - if you spot somewhere else let me know.

Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:19]: Will do.

just fyi. Something else came up, will continue the review tomorrow

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 3]
```diff
@@ -292,6 +304,22 @@ func statusVariableFromClusterClassVariable(variable clusterv1.ClusterClassVaria
...
+func addDefinitiontoExistingStatusVariable(variable clusterv1.ClusterClassVariable, from string, existingVariable clusterv1.ClusterClassStatusVariable) clusterv1.ClusterClassStatusVariable {
+	combinedVariable := existingVariable.DeepCopy()
+	newVariableDefinition := clusterv1.ClusterClassStatusVariableDefinition{
+		From:     from,
+		Required: variable.Required,
+		Schema:   variable.Schema,
+	}
+	combinedVariable.Definitions = append(existingVariable.Definitions, newVariableDefinition)
+
+	// If the new definition is different from any existing definition, set DefinitionsConflict to true.
```
Comment (sbueringer) [inline internal/controllers/clusterclass/clusterclass_controller.go:316]: Regarding reflect.DeepEqual:

Q: Reading through the README in https://github.com/google/go-cmp I really wonder if we should use cmp.Diff to compare stuff instead of reflect.DeepEqual. (in test and prod code)

The comparision algorithm seems to be better and additionally we get the beneift of getting a result which tells us where the difference is. This would be super valuable for our prod code to tell folks why we are doing things (using cmp.Diff for that while we use reflect.DeepEqual for the actual comparison is not great as there are small differences, e.g. unexported fields).

One example where I think this is super valuable is in the KCP controller where we could tell folks why exactly we are triggering a MachineRollout and not just: well deepequal told us that there is a difference.

Also here we could have a log message (with high v) where we could log where exactly the definitions are different.

I'm 100% fine with just using reflect.DeepEqual here like we do in (all?) other places today. But I think it would be valuable enough to open an issue to discuss migrating over. (I would take over opening that issue)

What is your opinion?

File: internal/webhooks/cluster_test.go [hunk 4]
```diff
@@ -60,245 +62,281 @@ func TestClusterDefaultNamespaces(t *testing.T) {
...
-			name: "default a single variable to its correct values",
-			clusterClass: builder.ClusterClass(metav1.NamespaceDefault, "class1").
-				WithVariables(clusterv1.ClusterClassVariable{
-					Name:     "location",
-					Required: true,
-					Schema: clusterv1.VariableSchema{
-						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
-							Type:    "string",
-							Default: &apiextensionsv1.JSON{Raw: []byte(`"us-east"`)},
+	var (
```
Comment (sbueringer) [inline internal/webhooks/cluster_test.go:65]: nit: Just curious, why did you move this into a var definition like this?

File: internal/webhooks/cluster_test.go [hunk 5]
```diff
@@ -320,11 +361,9 @@ func TestClusterDefaultVariables(t *testing.T) {
 			// Test if defaulting works in combination with validation.
 			util.CustomDefaultValidateTest(ctx, cluster, webhook)(t)
 			// Test defaulting.
-			t.Run("default", func(t *testing.T) {
```
Comment (sbueringer) [inline internal/webhooks/cluster_test.go:323]: nit: This was intentional as `util.CustomDefaultValidateTest` also creates sub-test-cases (named `validate-on-create`, ...)

Is there a reason why we can't keep it?

**Review by sbueringer — COMMENTED** [review]
Next iteration. And now reviewing internal/topology/variables

File: internal/controllers/clusterclass/clusterclass_controller.go (hunk 3 above)
Comment (killianmuldoon) [reply internal/controllers/clusterclass/clusterclass_controller.go:316]: Sounds like a good idea to me not sure if there's other alternatives to go-cmp we should consider. I think the right approach is to keep this in line with what we do elsewhere in CAPI and open an issue do discuss replacing it wholesale.

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 6]
```diff
@@ -30,66 +30,89 @@ import (
...
+func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassStatusVariable, createVariables bool, fldPath *field.Path) ([]clusterv1.ClusterVariable, field.ErrorList) {
 	var allErrs field.ErrorList
 
-	// Build maps for easier and faster access.
-	clusterVariablesMap := getClusterVariablesMap(clusterVariables)
-	clusterClassVariablesMap := getClusterClassVariablesMap(clusterClassVariables)
-
-	// Validate that all variables in the Cluster are defined in the ClusterClass.
-	// Note: If we don't validate this, we would get a nil pointer dereference below.
-	allErrs = append(allErrs, validateClusterVariablesDefined(clusterVariables, clusterClassVariablesMap, fldPath)...)
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:53]: Is it okay to just drop this? (or do we have a replacement for it that I missed?)

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 7]
```diff
@@ -30,66 +30,89 @@ import (
...
+				}
+				// Continue if the Cluster variable with an empty definitionFrom exists. The user intention here is to
+				// use the default value from a ClusterClass variable with no conflicting variables.
+				if _, ok := clusterVariablesMap[variable.Name][emptyVariableDefinitionFrom]; ok {
+					continue
+				}
+			}
+			allVariables = append(allVariables, clusterv1.ClusterVariable{
+				Name:           variable.Name,
+				DefinitionFrom: definition.From,
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:78]: Q: Doesn't this mean that we never add a global variable here? (as definition.From will always have some value)

Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting.go:78]: Also seems that when a variable with DefinitionsConflicts: false has to be entirely defauled we will default a value for each DefinitionFrom?

(these sentences would be really easier to read if I could write namespace :joy:)

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 8]
```diff
@@ -30,66 +30,89 @@ import (
...
+	for i, variable := range allVariables {
+		clusterClassVariable, found := clusterClassVariablesMap[ccVariableMapKey{name: variable.Name, from: variable.DefinitionFrom}]
+		if !found {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("name"), variable.Name,
+				fmt.Sprintf("variable with name %q from %q is not defined in ClusterClass `status.Variables`.", variable.Name, variable.DefinitionFrom)))
+			continue
+		}
 
-		defaultedClusterVariable, errs := defaultClusterVariable(clusterVariable, clusterClassVariable, fldPath.Index(i), createVariables)
+		// If the variable is defined in the Cluster spec get the definition.
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:93]: Didn't this happen in l.86?

File: internal/topology/variables/cluster_variable_defaulting.go (hunk 6 above)
Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting.go:53]: I assume because we don't get nil pointers anymore and then the variable validation will take care of this?

File: internal/topology/variables/cluster_variable_validation.go [hunk 9]
```diff
@@ -187,18 +224,61 @@ func validateUnknownFields(fldPath *field.Path, clusterVariable *clusterv1.Clust
...
-		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassStatusVariable) map[ccVariableMapKey]statusVariableDefinition {
+	variablesMap := map[ccVariableMapKey]statusVariableDefinition{}
+	for _, ccVar := range clusterClassVariables {
+		for _, varDef := range ccVar.Definitions {
+			variablesMap[ccVariableMapKey{name: ccVar.Name, from: varDef.From}] = statusVariableDefinition{
+				conflicts:                            ccVar.DefinitionsConflict,
+				ClusterClassStatusVariableDefinition: varDef,
+			}
+			if !ccVar.DefinitionsConflict {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation.go:272]: should this be one for loop up?

File: internal/topology/variables/cluster_variable_validation.go [hunk 10]
```diff
@@ -21,87 +21,124 @@ import (
...
+	for i, clusterVariable := range clusterVariables {
+		c := clusterVariable
+		// Check that the variable is defined in the ClusterClass `.status.variables`.
+		clusterClassVariable, found := clusterClassVariablesMap[ccVariableMapKey{clusterVariable.Name, clusterVariable.DefinitionFrom}]
+		if !found {
+			// This will catch cases where a user tries to define a variable without a definitionFrom field. In this
+			// case use a special error.
+			if clusterVariable.DefinitionFrom == emptyVariableDefinitionFrom {
+				allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("name"), clusterVariable.Name,
+					fmt.Sprintf("variable with name %q is not valid. It may not be defined in ClusterClass `status.Variables` or it may require `definitionFrom` in cluster `spec.variables`", clusterVariable.Name))) // TODO: consider if to add ClusterClass name
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation.go:88]: Would be nice for users if we can split this error up into two (aka differentiate the cases)

File: internal/topology/variables/cluster_variable_validation.go [hunk 11]
```diff
@@ -187,18 +224,61 @@ func validateUnknownFields(fldPath *field.Path, clusterVariable *clusterv1.Clust
...
+		return nil, kerrors.NewAggregate(errs)
+	}
+	return variablesMap, nil
 }
 
-func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassVariable) map[string]*clusterv1.ClusterClassVariable {
-	variablesMap := map[string]*clusterv1.ClusterClassVariable{}
-	for i := range clusterClassVariables {
-		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassStatusVariable) map[ccVariableMapKey]statusVariableDefinition {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation.go:264]: Wondering if we should return the same data structure as above in getClusterVariablesMap => map[variable-name][definitionfrom]

File: internal/topology/variables/cluster_variable_defaulting.go (hunk 6 above)
Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_defaulting.go:53]: The check for whether a variable is defined is in L:89 - or that's the intention

File: internal/topology/variables/cluster_variable_defaulting.go (hunk 8 above)
Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_defaulting.go:93]: That's where the ClusterClass variable comes from this is the ClusterVariable

File: internal/topology/variables/cluster_variable_validation.go (hunk 9 above)
Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_validation.go:272]: Easier here IMO as we need the variable definition.

**Comment by killianmuldoon** [comment]
Should be ready for another round @sbueringer

**Comment by sbueringer** [comment]
/retest

Some new flakes but independent of this PR (I'll probably take a look later)

File: internal/topology/variables/cluster_variable_validation.go [hunk 12]
```diff
@@ -21,87 +21,107 @@ import (
...
-				fmt.Sprintf("variable with name %q is not defined in ClusterClass `status.variables`", clusterVariable.Name))} // TODO: consider if to add ClusterClass name
+	for name, perPatchDefs := range definitions {
+		for _, def := range perPatchDefs {
+			// Check the required value for the specific variable definition. If the variable is not required continue.
+			if !def.Required {
+				continue
+			}
+
+			// If there is no variable with this name defined in the Cluster add an error and continue.
+			valuesForName, found := values[name]
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation.go:105]: Is this valuesByDefinitionFrom?

File: internal/topology/variables/cluster_variable_defaulting_test.go [hunk 13]
```diff
@@ -222,6 +317,118 @@ func Test_DefaultClusterVariables(t *testing.T) {
...
+				{
+					Name: "cpu",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`2`),
+					},
+				},
+			},
+		},
+		{
+			name: "Default non-specified definitions if only some are specified with definitionFrom",
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting_test.go:380]:
```suggestion
			name: "Default non-conflicting definitions if only some are specified with definitionFrom",
```
?

File: internal/topology/variables/cluster_variable_validation.go [hunk 14]
```diff
@@ -21,87 +21,107 @@ import (
...
 
-// ValidateClusterVariables validates ClusterVariables.
-func ValidateClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
-	return validateClusterVariables(clusterVariables, clusterClassVariables, true, fldPath)
+const (
+	// emptyVariableDefinitionFrom is the definitionFrom value used when none is supplied by the user.
+	emptyVariableDefinitionFrom = ""
+)
+
+type statusVariableDefinition struct {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/topology/variables/cluster_variable_validation.go:40]: Let's use upper-case for all fields names. It's enough that the struct is unexported

File: internal/topology/variables/cluster_variable_defaulting_test.go [hunk 15]
```diff
@@ -315,13 +529,16 @@ func Test_DefaultClusterVariable(t *testing.T) {
...
-				Name:     "location",
-				Required: true,
-				Schema: clusterv1.VariableSchema{
-					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
-						Type:    "string",
-						Default: &apiextensionsv1.JSON{Raw: []byte(`"us-east"`)},
+			clusterClassVariable: &statusVariableDefinition{
+				name: "location",
+				ClusterClassStatusVariableDefinition: &clusterv1.ClusterClassStatusVariableDefinition{
+
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting_test.go:535]: nit: let's drop the new lines that we have in a few cases

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 16]
```diff
@@ -27,215 +27,6 @@ import (
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 )
 
-func Test_ValidateClusterVariables(t *testing.T) {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation_test.go:30]: Let's please keep the test order the same for this PR. It just makes the diff unnecessary hard 
(also not sure why the order was changed)

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 17]
```diff
@@ -1133,3 +924,666 @@ func Test_ValidateClusterVariable(t *testing.T) {
...
+								},
+							},
+						},
+					},
+				},
+				{
+					Name: "zone",
+					Definitions: []clusterv1.ClusterClassStatusVariableDefinition{
+						{
+
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation_test.go:958]: nit: same here, can we drop the empty lines?

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 18]
```diff
@@ -1133,3 +924,666 @@ func Test_ValidateClusterVariable(t *testing.T) {
...
+			},
+			validateRequired: true,
+		},
+		{
+			name:    "Fail if variable DefinitionFrom field does not match an existing definition.",
+			wantErr: true,
+			clusterClassVariables: []clusterv1.ClusterClassStatusVariable{
+				{
+					Name: "cpu",
+					// There are conflicting definitions which means Cluster Variables should include a `from` field.
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation_test.go:1212]: comment seems to be wrong here. I think we can just drop it, name is good enough for me here

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 19]
```diff
@@ -1133,3 +924,666 @@ func Test_ValidateClusterVariable(t *testing.T) {
...
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`1`),
+					},
+				},
+			},
+			validateRequired: true,
+		},
+		{
+			name:    "Fail if variable DefinitionFrom field does not match an existing definition.",
+			wantErr: true,
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation_test.go:1208]: nit: let's set wantErr either consistently at the beginning or end of the test case struct

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 20]
```diff
@@ -1133,3 +924,666 @@ func Test_ValidateClusterVariable(t *testing.T) {
...
+			validateRequired: true,
+		},
+		{
+			name:    "Fail if a required variable with conflicting definitions isn't defined for each separate DefinitionFrom.",
+			wantErr: true,
+			clusterClassVariables: []clusterv1.ClusterClassStatusVariable{
+				{
+					Name:                "cpu",
+					DefinitionsConflict: true,
+					// There are conflicting definitions which means Cluster Variables should include a `from` field.
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation_test.go:1243]:
```suggestion
```

File: internal/topology/variables/cluster_variable_validation_test.go [hunk 21]
```diff
@@ -1133,3 +924,666 @@ func Test_ValidateClusterVariable(t *testing.T) {
...
+									Type: "integer",
+								},
+							},
+							From: "somepatch",
+						},
+					},
+				},
+			},
+			clusterVariables: []clusterv1.ClusterVariable{
+				// Note: this is a different case than the above because of the ordering of the variables.
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation_test.go:1382]: I think this is not correct. It seems to be an entirely different case

File: internal/topology/variables/cluster_variable_validation.go (hunk 12 above)
Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_validation.go:105]: This is all variable values with a given name

Comment (sbueringer) [reply internal/topology/variables/cluster_variable_validation.go:99]: Ah I see, yeah makes sense to think this way around :)

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 22]
```diff
@@ -213,15 +215,13 @@ func (r *Reconciler) reconcileExternalReferences(ctx context.Context, clusterCla
...
+	allVariableDefinitions := map[string]*clusterv1.ClusterClassStatusVariable{}
 	// Add inline variable definitions to the ClusterClass status.
 	for _, variable := range clusterClass.Spec.Variables {
-		uniqueVars[variable.Name] = true
-		allVars = append(allVars, statusVariableFromClusterClassVariable(variable, clusterv1.VariableDefinitionFromInline))
+		allVariableDefinitions[variable.Name] = addNewStatusVariable(variable, clusterv1.VariableDefinitionFromInline)
 	}
 
-	// If RuntimeSDK is enabled call the DiscoverVariables hook for all associated Runtime Extensions and add the variables
+	// If RuntimeSDK is enabled call the DiscoverVariables hook for all associated Runtime Extensions and add the patchResponse
```
Comment (sbueringer) [inline internal/controllers/clusterclass/clusterclass_controller.go:224]: nit: Should this stay "variables"? (same in l.267 and l.270)

File: internal/controllers/clusterclass/clusterclass_controller_test.go [hunk 23]
```diff
@@ -308,3 +315,231 @@ func isOwnerReferenceEqual(a, b metav1.OwnerReference) bool {
...
+				Build()
+
+			r := &Reconciler{
+				RuntimeClient: fakeRuntimeClient,
+			}
+
+			err := r.reconcileVariables(ctx, tt.clusterClass)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+			}
```
Comment (sbueringer) [inline internal/controllers/clusterclass/clusterclass_controller_test.go:541]: nit: let's also check that it hasn't occured in the else branch

File: internal/controllers/clusterclass/clusterclass_controller_test.go [hunk 24]
```diff
@@ -308,3 +315,231 @@ func isOwnerReferenceEqual(a, b metav1.OwnerReference) bool {
 	}
 	return true
 }
+
+func TestReconciler_reconcileVariables(t *testing.T) {
```
Comment (sbueringer) [inline internal/controllers/clusterclass/clusterclass_controller_test.go:319]: **Praise**: nice test! Concise, easy to read and covers all relevant cases!

**Review by sbueringer — COMMENTED** [review]
Reviewed internal/controllers and internal/webhooks

Now continuing with internal/topology/variables

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 25]
```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
...
+// getCurrentValue returns the value of a variable for its definitionFrom, or for an empty definitionFrom if it exists.
+func getCurrentValue(variable clusterv1.ClusterVariable, valuesMap map[string]map[string]clusterv1.ClusterVariable) *clusterv1.ClusterVariable {
+	// If the value is set in the Cluster spec get the value.
+	if valuesForName, ok := valuesMap[variable.Name]; ok {
+		if value, ok := valuesForName[variable.DefinitionFrom]; ok {
+			return &value
+		}
+		if value, ok := valuesForName[emptyDefinitionFrom]; ok {
+			return &value
+		}
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:105]: Wondering if we need / should have that fallback. Could it be that DefinitionFrom is already either set to an empty or non-empty value and thus we don't need the fallback here?

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 26]
```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
...
 
 	var value interface{}
 	// If the variable already exists, parse the current value.
-	if clusterVariable != nil && len(clusterVariable.Value.Raw) > 0 {
-		if err := json.Unmarshal(clusterVariable.Value.Raw, &value); err != nil {
+	if currentValue != nil && len(currentValue.Value.Raw) > 0 {
+		if err := json.Unmarshal(currentValue.Value.Raw, &value); err != nil {
 			return nil, field.ErrorList{field.Invalid(fldPath, "",
-				fmt.Sprintf("failed to unmarshal variable value %q: %v", string(clusterVariable.Value.Raw), err))}
+				fmt.Sprintf("failed to unmarshal variable %q value %q: %v", definition.Name, string(currentValue.Value.Raw), err))}
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:135]:
```suggestion
				fmt.Sprintf("failed to unmarshal variable %q value %q: %v", currentValue.Name, string(currentValue.Value.Raw), err))}
```
nit: I know they are the same but using the value name seems more intuitive (given that the entire if block is based on currentValue)

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 27]
```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
...
+			if _, ok := valuesIndex[variable.Name]; ok {
+				if _, ok := valuesIndex[variable.Name][definitionFrom]; ok {
+					continue
+				}
+				if _, ok := valuesIndex[variable.Name][emptyDefinitionFrom]; ok {
+					continue
+				}
+			}
+
+			// 2) If the definition has no conflicts no variable of the same name is defined in the Cluster set the definitionFrom to emptyDefinitionFrom.
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:204]:
```suggestion
			// 2) If the definition has no conflicts and no variable of the same name is defined in the Cluster set the definitionFrom to emptyDefinitionFrom.
```
nit

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 28]
```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
...
+			if !variable.DefinitionsConflict && len(valuesIndex[variable.Name]) == 0 {
+				definitionFrom = emptyDefinitionFrom
+			}
+
+			// 3) If a variable with this name and definition has been added already, continue.
+			if _, ok := uniqueVariableDefinitions[definitionFrom+variable.Name]; ok {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:210]: Let's add a godoc comment in which case we are hitting this if
(I think it's the case where we have multiple definitions with no conflict and no value in the Cluster spec, which is roughly when we hit case 2. in a scenario where we have multiple definitions (?))

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 29]
```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
...
 		},
-	}, nil
+		DefinitionFrom: definition.From,
+	}
+
+	return v, nil
+}
+
+// getAllVariables returns a correctly ordered list of all variables defined in the ClusterClass and the Cluster.
+func getAllVariables(values []clusterv1.ClusterVariable, valuesIndex map[string]map[string]clusterv1.ClusterVariable, definitions []clusterv1.ClusterClassStatusVariable) []clusterv1.ClusterVariable {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:181]: I think it would be good to have a unit test at this level to cover all the permutations

**Review by sbueringer — COMMENTED** [review]
Reviewed prod code in internal/topology/variables

Great work!

Almost only nits. Now onto unit tests

File: internal/topology/variables/cluster_variable_defaulting_test.go [hunk 30]
```diff
@@ -29,15 +29,15 @@ import (
 func Test_DefaultClusterVariables(t *testing.T) {
 	tests := []struct {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting_test.go:30]: Just something to think about:

Do we have test coverage (either in defaulting or validation) for the cases where we "transition" between the various ways variables can be / have to be defined:
* 1 global variable
* namespaced variables because of conflict
* namespaced variables because users chose too

With transition I mean that the Cluster.spec.topology variables is using one of those options and based on variable definitions we default/validate based on another one. (for all permutations)

Could be that we already have all permutations and a lot of those are probably just error cases. Would be great if you can go through the permutations and double check that we have them all covered.

File: internal/topology/variables/cluster_variable_defaulting_test.go [hunk 31]
```diff
@@ -247,20 +505,23 @@ func Test_DefaultClusterVariable(t *testing.T) {
...
 		name                 string
 		clusterVariable      *clusterv1.ClusterVariable
-		clusterClassVariable *clusterv1.ClusterClassVariable
+		clusterClassVariable *statusVariableDefinition
+		conflict             bool
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting_test.go:509]:
```suggestion
```
nit: unused

File: internal/topology/variables/cluster_variable_defaulting_test.go (hunk 30 above)
Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting_test.go:30]: It would be great if we can order / group the test cases in a way that's easier to assess if we covered all cases (+ maybe add a quick godoc comment over each section) (starting here as everything above are the straightforward cases)

E.g. One group for each of
* conflicting definitions
* non-conflicting definitions
(not sure if that's everything)

And then go through each group for the various variants how the Cluster.spec.topology.variables could look like:
* no variable defined at all
* a global one defined
* a namespaced one defined
* all namespaced ones defined (no-op)
* ...


Would be nice if we can also write the names in a consistent way. Currently it's not easy to figure out which cases we cover and if there are gaps

(same for validation tests)

Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting_test.go:30]: One additional point. Do we have enough coverage for cases where for one variable name some definitions are optional and others are required? (for defaulting and validation)

File: internal/topology/variables/cluster_variable_defaulting.go (hunk 25 above)
Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting.go:107]: I think I realized later that we need it. But please double check

Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_defaulting.go:107]: I think you're right with the changes in getAllVariables.

**Comment by sbueringer** [comment]
Sorry my bad. Have to fix the job config..

File: internal/topology/variables/cluster_variable_defaulting_test.go (hunk 30 above)
Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_defaulting_test.go:30]: >With transition I mean that the Cluster.spec.topology variables is using one of those options and based on variable definitions we default/validate based on another one. (for all permutations)

Not sure I understand what the ask is here. I'm currently cleaning up the tests to be more ordered. Let's sync on whether they're good enough once I've pushed that change (+/- time to review)

File: internal/topology/variables/utils.go [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+
+	// If definitionFrom is empty and there are no conflicts return a definition with an emptyDefinitionFrom.
+	if definitionFrom == emptyDefinitionFrom {
+		for _, def := range i[name] {
+			if !def.Conflicts {
+				return &statusVariableDefinition{
+					Name:      def.Name,
+					Conflicts: def.Conflicts,
+					ClusterClassStatusVariableDefinition: &clusterv1.ClusterClassStatusVariableDefinition{
+						// Return the definition with an empty definitionFrom. This ensures when a user get s
```
Comment (sbueringer) [inline internal/topology/variables/utils.go:135]:
```suggestion
						// Return the definition with an empty definitionFrom. This ensures when a user gets
```
nit

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 33]
```diff
@@ -242,14 +242,23 @@ func (r *Reconciler) reconcileVariables(ctx context.Context, clusterClass *clust
 				continue
 			}
 			if resp.Variables != nil {
+				uniqueNamesForPatch := map[string]bool{}
```
Comment (fabriziopandini) [inline internal/controllers/clusterclass/clusterclass_controller.go:245]: nit (feel free to ignore)
```suggestion
				variableNames := sets.Set[string]()
```
so below you can uset variableNames.Has and variableNames.Insert which reads better

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 34]
```diff
@@ -260,10 +269,20 @@ func (r *Reconciler) reconcileVariables(ctx context.Context, clusterClass *clust
 			"VariableDiscovery failed: %s", kerrors.NewAggregate(errs))
 		return errors.Wrapf(kerrors.NewAggregate(errs), "failed to discover variables for ClusterClass %s", clusterClass.Name)
 	}
-	clusterClass.Status.Variables = allVars
+
+	statusVarList := []clusterv1.ClusterClassStatusVariable{}
+	for _, variable := range allVariableDefinitions {
+		statusVarList = append(statusVarList, *variable)
+	}
+	// Alphabetically sort the variables by name. This ensures no unnecessary updates to the ClusterClass status.
```
Comment (fabriziopandini) [inline internal/controllers/clusterclass/clusterclass_controller.go:277]: If I'm not wrong, if we want to achieve stable ordering we must sort also definitions nested in each entry; this can be done both here (preferred) or where we add definitions

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 35]
```diff
@@ -86,83 +83,141 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
...
+		}
+	}
+	return nil
 }
 
-// defaultClusterVariable defaults a clusterVariable based on the default value in the clusterClassVariable.
-func defaultClusterVariable(clusterVariable *clusterv1.ClusterVariable, clusterClassVariable *clusterv1.ClusterClassVariable, fldPath *field.Path, createVariable bool) (*clusterv1.ClusterVariable, field.ErrorList) {
-	if clusterVariable == nil {
+// defaultValue defaults a clusterVariable based on the default value in the clusterClassVariable.
+func defaultValue(currentValue *clusterv1.ClusterVariable, definition *statusVariableDefinition, fldPath *field.Path, createVariable bool) (*clusterv1.ClusterVariable, field.ErrorList) {
```
Comment (fabriziopandini) [inline internal/topology/variables/cluster_variable_defaulting.go:110]: kudos for cleaning up the wording (the distinction between value/definition make code more readable)

**Review by fabriziopandini — COMMENTED** [review]
This is outstanding work!
Just few comments from my side

File: internal/controllers/clusterclass/clusterclass_controller.go (hunk 34 above)
Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:277]: +1 It's basically the same as the order in which patches are applied is always stable because we always just iterate through the array in the ClusterClass.

File: internal/topology/variables/cluster_variable_defaulting.go (hunk 35 above)
Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting.go:110]: Absolutely agree. clusterVariable vs clusterClassVariable was driving me slightly crazy :)
(disclaimer I wrote this code initially)

File: internal/controllers/clusterclass/clusterclass_controller.go (hunk 34 above)
Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:277]: Maybe let's add a short note about that this part is already stable here?

**Comment by killianmuldoon** [comment]
/hold remove

Underlying PRs all merged

**Review by sbueringer — COMMENTED** [review]
Last nits from my side

**Comment by sbueringer** [comment]
Thank you very much!!

Really great work, thank you especially for the extensive test coverage!!!

/lgtm

**Comment by fabriziopandini** [comment]
yay!
/lgtm
/approve
